crates/dbsp/src/operator/dynamic/balance/balancer.rs

```diff
@@ -0,0 +1,832 @@
+    joins: BTreeMap<NodeId, (NodeId, NodeId, bool)>,
+
+    /// Connected components of the join graph:
+    clusters: Vec<Cluster>,
+
+    /// Map from input stream to the cluster index it belongs to.
+    stream_to_cluster: BTreeMap<NodeId, usize>,
+
+    /// Circuit's metadata exchange used to collect metadata from RebalancingExchangeSender instances.
+    metadata_exchange: MetadataExchange,
+}
+
+impl BalancerInner {
+    fn new(metadata_exchange: &MetadataExchange) -> Self {
+        Self {
+            integrals: BTreeMap::new(),
+            joins: BTreeMap::new(),
+            clusters: Vec::new(),
+            stream_to_cluster: BTreeMap::new(),
+            metadata_exchange: metadata_exchange.clone(),
+        }
+    }
+
+    /// Compute an "optimal" policy for each stream in the cluster.
+    fn solve_cluster(
+        &self,
+        cluster_index: usize,
+    ) -> Result<BTreeMap<NodeId, PartitioningPolicy>, BalancerError> {
+        //let cluster = &mut self.clusters[cluster_index];
+
+        let mut maxsat = MaxSat::new();
+        let mut variable_indexes = BTreeMap::new();
+
+        let mut domains = BTreeMap::new();
+
+        let mut current_effective_policy = BTreeMap::new();
+
+        // Create a MaxSat variable for each stream in the cluster.
+        for stream in self.clusters[cluster_index].streams.clone().iter() {
+            let (exchange_sender, hints) = self.integrals.get(stream).unwrap();
+
+            let exchange_sender_metadata: Vec<Option<RebalancingExchangeSenderExchangeMetadata>> = self
+                .metadata_exchange
+                .get_global_operator_metadata_typed::<RebalancingExchangeSenderExchangeMetadata>(
+                    *exchange_sender,
+                );
+
+            if let Some(metadata) = &exchange_sender_metadata
+                .first()
+                .and_then(|metadata| metadata.as_ref())
+            {
+                current_effective_policy.insert(*stream, metadata.current_policy);
+            }
+
+            let fixed_policy = RebalancingExchangeSenderExchangeMetadata::get_fixed_policy(
+                &exchange_sender_metadata,
+                hints,
+            )?;
+
+            // println!(
+            //     "{} fixed_policy for {stream} (hints: {hints:?}): {:?}",
+            //     Runtime::worker_index(),
+            //     fixed_policy,
+            // );
+
+            let mut domain = self.compute_domain(&exchange_sender_metadata, hints);
+            if let Some(fixed_policy) = &fixed_policy {
+                domain.retain(|policy, _| *policy == *fixed_policy);
+            }
+
+            let variable_index = maxsat.add_variable(&stream.to_string(), &domain);
+            variable_indexes.insert(*stream, variable_index);
+            domains.insert(*stream, domain);
+        }
+
+        for (_join, (left, right, is_left_join)) in self.clusters[cluster_index].joins.iter() {
+            maxsat.add_constraint(JoinConstraint::new(
```

Comment:
these constraints are the same always so you can just build them once and clone them
but maybe that is overoptimizing